rescan.go

```diff
@@ -885,127 +885,32 @@ func (s *ChainService) GetUtxo(options ...RescanOption) (*SpendReport, error) {
-		} else if ro.startBlock.Height < int32(curHeight) {
-			header, err := s.BlockHeaders.FetchHeaderByHeight(
-				uint32(ro.startBlock.Height))
-			if err == nil {
-				ro.startBlock.Hash = header.BlockHash()
-			} else {
-				ro.startBlock.Hash = *s.chainParams.GenesisHash
-				ro.startBlock.Height = 0
-			}
-		} else {
-			ro.startBlock.Height = int32(curHeight)
-			ro.startBlock.Hash = curHeader.BlockHash()
-		}
-	}
-	log.Tracef("Starting scan for output spend from known block %d (%s) "+
-		"back to block %d (%s)", curStamp.Height, curStamp.Hash,
-		ro.startBlock.Height, ro.startBlock.Hash)
-
-	for {
-		// Check the basic filter for the spend and the extended filter
-		// for the transaction in which the outpoint is funded.
-		filter, err := s.GetCFilter(curStamp.Hash,
-			wire.GCSFilterRegular, ro.queryOptions...)
-		if err != nil {
-			return nil, fmt.Errorf("Couldn't get basic "+
-				"filter for block %d (%s)", curStamp.Height,
-				curStamp.Hash)
-		}
-		matched := false
-		if filter != nil {
-			filterKey := builder.DeriveKey(&curStamp.Hash)
-			matched, err = filter.MatchAny(filterKey, watchList)
-		}
-		if err != nil {
-			return nil, err
-		}
-
-		// If either is matched, download the block and check to see
-		// what we have.
-		if matched {
-			block, err := s.GetBlockFromNetwork(curStamp.Hash,
-				ro.queryOptions...)
-			if err != nil {
-				return nil, err
-			}
-			if block == nil {
-				return nil, fmt.Errorf("Couldn't get "+
-					"block %d (%s)", curStamp.Height,
-					curStamp.Hash)
-			}
-
-			// If we've spent the output in this block, return an
-			// error stating that the output is spent.
-			for _, tx := range block.Transactions() {
-				for i, ti := range tx.MsgTx().TxIn {
-					if ti.PreviousOutPoint == ro.watchOutPoints[0] {
-						return &SpendReport{
-							SpendingTx:         tx.MsgTx(),
-							SpendingInputIndex: uint32(i),
-							SpendingTxHeight:   uint32(curStamp.Height),
-						}, nil
-					}
-				}
-			}
-
-			// If we found the transaction that created the output,
-			// then it's not spent and we can return the TxOut.
-			for _, tx := range block.Transactions() {
-				if *(tx.Hash()) == ro.watchOutPoints[0].Hash {
-					outputs := tx.MsgTx().TxOut
-					return &SpendReport{
-						Output: outputs[ro.watchOutPoints[0].Index],
-					}, nil
-				}
-			}
-		}
 
-		// Otherwise, iterate backwards until we've gone too far.
-		curStamp.Height--
-		if curStamp.Height < ro.startBlock.Height {
-			return nil, fmt.Errorf("Transaction %s not found "+
-				"since start block %d (%s)",
-				ro.watchOutPoints[0].Hash, curStamp.Height+1,
-				curStamp.Hash)
-		}
+	var (
+		result *SpendReport
+		resultError error
+		wg sync.WaitGroup
+	)
 
-		// Fetch the previous header so we can continue our walk
-		// backwards.
-		header, err := s.BlockHeaders.FetchHeaderByHeight(uint32(curStamp.Height))
-		if err != nil {
-			return nil, err
-		}
-		curStamp.Hash = header.BlockHash()
-	}
+	wg.Add(1)
```

Comment:
This very much relies on the code 100% of the time executing the Result function. Which it seems like it does, but that promise is easily forgotten/broken. I would have gone for a `chan struct{}` that is closed in the result function, so that you can abort waiting when the service is closing:
```go
select {
case <-s.quit:
	return nil, errors.New("service shutting down")
case <-resultChan:
}
```